File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/EventAnnouncer.java
package com.gmail.borlandlp.minigamesdtools.arena;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.arena.localevent.*;
import org.bukkit.event.Cancellable;

import java.lang.reflect.Method;
import java.util.*;

public class EventAnnouncer {
    private Map<Class<? extends ArenaEvent>, List<RegisteredArenaListener>> listeners = new HashMap<>();

    public void register(ArenaEventListener listener) {
        for (Object IMethod : getObjectMethods(listener)) {
            ArenaEventHandler annClass = ((Method)IMethod).getAnnotation(ArenaEventHandler.class);
            if(((Method) IMethod).getGenericParameterTypes().length == 1 && annClass != null) {
                Class argEventClazz = null;

                try {
                    argEventClazz = Class.forName((((Method)IMethod).getGenericParameterTypes()[0]).getTypeName());
                } catch (ClassNotFoundException e) {
                    e.printStackTrace();
                }
                if(argEventClazz == null) {
                    String parameterClass = (((Method)IMethod).getGenericParameterTypes()[0]).getTypeName();
                    String listenerClass = listener.getClass().getSimpleName();
                    String msg = "Internal error. Cant found class " + parameterClass + " for " + listenerClass;
                    Debug.print(Debug.LEVEL.WARNING, msg);
                    continue;
                }

                registerListener(argEventClazz, listener, annClass, (Method) IMethod);
            }
        }
    }

    private void registerListener(Class argEventClazz, ArenaEventListener listener, ArenaEventHandler annClass, Method method) {
        List<RegisteredArenaListener> handlers = this.getListenersOf(argEventClazz);
        RegisteredArenaListener rListener = new RegisteredArenaListener(listener, method, annClass.ignoreCancelled(), annClass.priority());
        handlers.add(rListener);
        this.listeners.put(argEventClazz, handlers);
    }

    private Set<Method> getObjectMethods(Object obj) {
        Method[] publicMethods = obj.getClass().getMethods();
        Method[] privateMethods = obj.getClass().getDeclaredMethods();
        Set<Method> methods = new HashSet<>(publicMethods.length + privateMethods.length, 1.0F);

        methods.addAll(Arrays.asList(publicMethods));
        methods.addAll(Arrays.asList(privateMethods));

        return methods;
    }

    public void unregister(ArenaEventListener listener) {
        for (List<RegisteredArenaListener> list : this.getListeners().values()) {
            list.removeIf(AListener -> AListener.getHandler() == listener);
        }
    }

    public Map<Class<? extends ArenaEvent>, List<RegisteredArenaListener>> getListeners() {
        return this.listeners;
    }

    public List<RegisteredArenaListener> getListenersOf(Class<? extends ArenaEvent> clazz) {
        return this.getListeners().getOrDefault(clazz, new ArrayList<>());
    }

    public <T extends ArenaEvent> void announce(T event) {
        if(!this.getListeners().containsKey(event.getClass())) {
            return;
        }

        for (ArenaEventPriority priority : ArenaEventPriority.values()) { // highest, high, normal, low, lowest...
            for(RegisteredArenaListener registeredArenaListener : this.getListenersOf(event.getClass())) {

                boolean isValidPriority = registeredArenaListener.getPriority() == priority;
                boolean hasCanceled = event instanceof Cancellable && ((Cancellable) event).isCancelled();
                boolean isCanceled = hasCanceled && !registeredArenaListener.isIgnoreCanceled();
                if(!isValidPriority || isCanceled) {
                    continue;
                }

                try {
                    registeredArenaListener.execute(event);
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/EventAnnouncer.kt
package com.gmail.borlandlp.minigamesdtools.arena

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaEvent
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaEventHandler
import com.gmail.borlandlp.minigamesdtools.arena.localevent.ArenaEventPriority
import org.bukkit.event.Cancellable
import java.lang.reflect.Method
import java.util.*

class EventAnnouncer {
    private val listeners: MutableMap<Class<out ArenaEvent>, MutableList<RegisteredArenaListener>> =
        HashMap()

    fun register(listener: ArenaEventListener) {
        for (IMethod in getObjectMethods(listener)) {
            val annClass =
                IMethod.getAnnotation(ArenaEventHandler::class.java)
            if (IMethod.genericParameterTypes.size == 1 && annClass != null) {
                val argEventClazz: Class<out ArenaEvent>? = try {
                     Class.forName(IMethod.genericParameterTypes[0].typeName) as Class<out ArenaEvent>?
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                    null
                }
                if (argEventClazz == null) {
                    val parameterClass =
                        IMethod.genericParameterTypes[0].typeName
                    val listenerClass = listener.javaClass.simpleName
                    val msg =
                        "Internal error. Cant found class $parameterClass for $listenerClass"
                    print(
                        Debug.LEVEL.WARNING,
                        msg
                    )
                    continue
                }
                registerListener(argEventClazz, listener, annClass, IMethod)
            }
        }
    }

    private fun registerListener(
        argEventClazz: Class<out ArenaEvent>,
        listener: ArenaEventListener,
        annClass: ArenaEventHandler,
        method: Method
    ) {
        val handlers = getListenersOf(argEventClazz)
        val rListener =
            RegisteredArenaListener(listener, method, annClass.ignoreCancelled, annClass.priority)
        handlers.add(rListener)
        listeners[argEventClazz] = handlers
    }

    private fun getObjectMethods(obj: Any): Set<Method> {
        val publicMethods = obj.javaClass.methods
        val privateMethods = obj.javaClass.declaredMethods
        val methods: MutableSet<Method> =
            HashSet(publicMethods.size + privateMethods.size, 1.0f)
        methods.addAll(Arrays.asList(*publicMethods))
        methods.addAll(Arrays.asList(*privateMethods))
        return methods
    }

    fun unregister(listenerToRemove: ArenaEventListener) {
        for (list in getListeners().values) {
            list.removeIf { listener ->
                listener.handler === listenerToRemove
            }
        }
    }

    fun getListeners(): Map<Class<out ArenaEvent>, MutableList<RegisteredArenaListener>> {
        return listeners
    }

    fun getListenersOf(clazz: Class<out ArenaEvent>): MutableList<RegisteredArenaListener> {
        return getListeners().getOrDefault(clazz, ArrayList<RegisteredArenaListener>()).toMutableList()
    }

    fun <T : ArenaEvent> announce(event: T) {
        if (!getListeners().containsKey(event.javaClass)) {
            return
        }
        for (priority in ArenaEventPriority.values()) { // highest, high, normal, low, lowest...
            for (registeredArenaListener in getListenersOf(event.javaClass)) {
                val isValidPriority = registeredArenaListener.priority === priority
                val hasCanceled =
                    event is Cancellable && (event as Cancellable).isCancelled
                val isCanceled = hasCanceled && !registeredArenaListener.isIgnoreCanceled
                if (!isValidPriority || isCanceled) {
                    continue
                }
                try {
                    registeredArenaListener.execute(event)
                } catch (ex: Exception) {
                    ex.printStackTrace()
                }
            }
        }
    }
}